### PR TITLE
etcd-tester: integrate etcd runner into etcd tester

### LINKED ISSUE #7026: functional-tester: Integrate etcd-runner into etcd-tester
We need to make etcd-runner a long running test including fault injection like what we do for etcd-tester. The long running test is a great way to catch obscure bugs that might happening only after running for a very long time. We want to make sure that the client code is robust even after being running for a long time and is working as expected under various failures.

### PR DESCRIPTION
**Pull request by fanminshi**
etcd tester runs etcd runner as a separate binary.
it signals sigstop to the runner when tester wants to stop stressing.
it signals sigcont to the runner when tester wants to start stressing.
when tester needs to clean up, it signals sigint to runner.

FIXES #7026

### CODE REVIEW

File: pkg/stringutil/stringutil.go [hunk 1]
```diff
@@ -14,13 +14,20 @@
...
+	"math/rand"
+	"time"
+)
 
 const (
 	chars = "abcdefghijklmnopqrstuvwxyz0123456789"
 )
 
-// UniqueStrings retruns a slice of randomly generated unique strings.
+func init() {
```
Comment (heyitsanthony) [inline pkg/stringutil/stringutil.go]: This will override rand seed setting for anything that imports this package. It should the be importer's responsibility to seed rand, so don't to this in init please.

File: tools/functional-tester/etcd-runner/command/election_command.go [hunk 2]
```diff
@@ -49,6 +50,8 @@ func runElectionFunc(cmd *cobra.Command, args []string) {
 	eps := endpointsFromFlag(cmd)
 	dialTimeout := dialTimeoutFromCmd(cmd)
 
+	elector := stringutil.RandomStrings(10, 1)[0]
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/election_command.go]: s/elector/election since it's a name of a particular election. Electors are participants.

File: pkg/stringutil/stringutil.go [hunk 3]
```diff
@@ -14,13 +14,20 @@
...
+	rand.Seed(time.Now().UTC().UnixNano())
```
Comment (xiang90) [inline pkg/stringutil/stringutil.go]: can you put this into a separate pr?

File: tools/functional-tester/etcd-runner/command/election_command.go [hunk 4]
```diff
@@ -65,8 +68,8 @@ func runElectionFunc(cmd *cobra.Command, args []string) {
 				break
 			}
 		}
-		e := concurrency.NewElection(s, "electors")
 
+		e := concurrency.NewElection(s, elector)
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/election_command.go]: why does this need to change? is it because it's possible there's a conflict? If so, how does having a random string solve the conflict, instead of just making it rare?

File: tools/functional-tester/etcd-runner/command/lease_renewer_command.go [hunk 5]
```diff
@@ -54,7 +53,7 @@ func runLeaseRenewerFunc(cmd *cobra.Command, args []string) {
 			err error
 		)
 		for {
-			l, err = c.Lease.Grant(ctx, 5)
+			l, err = c.Lease.Grant(ctx, 30)
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/lease_renewer_command.go]: why does this need to be bumped up? maybe it should be a command line argument?

File: tools/functional-tester/etcd-runner/command/lock_racer_command.go [hunk 6]
```diff
@@ -61,7 +64,7 @@ func runRacerFunc(cmd *cobra.Command, args []string) {
 				break
 			}
 		}
-		m := concurrency.NewMutex(s, "racers")
+		m := concurrency.NewMutex(s, racers)
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/lock_racer_command.go]: same as election code-- if there's a conflict I don't think using a random string really solves the problem

File: pkg/stringutil/stringutil.go (hunk 3 above)
Comment (fanminshi) [inline pkg/stringutil/stringutil.go]: will do

File: tools/functional-tester/etcd-tester/stresser.go [hunk 7]
```diff
@@ -26,10 +26,12 @@ import (
 func init() { grpclog.SetLogger(plog) }
 
 type Stresser interface {
-	// Stress starts to stress the etcd cluster
-	Stress() error
-	// Cancel cancels the stress test on the etcd cluster
-	Cancel()
+	// Start starts to stress the etcd cluster
+	Start() error
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/stresser.go]: keep this as `Stress`

File: tools/functional-tester/etcd-tester/stresser.go [hunk 8]
```diff
@@ -26,10 +26,12 @@ import (
...
+	// Stop stops stressing on the etcd cluster
+	Stop()
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/stresser.go]: `Pause`?

File: pkg/stringutil/stringutil.go (hunk 3 above)
Comment (fanminshi) [inline pkg/stringutil/stringutil.go]: done via https://github.com/coreos/etcd/pull/7673

File: tools/functional-tester/etcd-tester/stresser.go [hunk 9]
```diff
@@ -149,6 +166,50 @@ func NewStresser(s string, sc *stressConfig, m *member) Stresser {
 			keysPerLease: sc.keysPerLease,
 			rateLimiter:  sc.rateLimiter,
 		}
+	case "election-runner":
+		return &runnerStresser{
+			cmdStr: sc.etcdRunnerPath,
+			args: []string{
+				"election",
+				"--rounds=10000000", // large enough to run a long time.
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/stresser.go]: I feel like etcd-runner should treat --rounds=0 as running forever instead of needing this hack

File: tools/functional-tester/etcd-tester/etcd_runner_stresser.go [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+	"syscall"
+)
+
+type runnerStresser struct {
+	cmd    *exec.Cmd
+	cmdStr string
+	args   []string
+
+	errorc chan error
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/etcd_runner_stresser.go]: `errc` is conventional

File: tools/functional-tester/etcd-tester/etcd_runner_stresser.go [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+func (rs *runnerStresser) setupOnce() (err error) {
+	if rs.cmdStr == "" {
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/etcd_runner_stresser.go]: probably doesn't need this error check, it should be caught by `cmd.Start()`

File: tools/functional-tester/etcd-tester/etcd_runner_stresser.go [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+func (rs *runnerStresser) Stop() {
+	syscall.Kill(rs.cmd.Process.Pid, syscall.SIGSTOP)
+}
+
+func (rs *runnerStresser) Close() {
+	syscall.Kill(rs.cmd.Process.Pid, syscall.SIGINT)
+	rs.cmd.Wait()
+}
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/etcd_runner_stresser.go]: <-rs.donec

File: tools/functional-tester/etcd-tester/etcd_runner_stresser.go [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+		return fmt.Errorf("expect cmdStr to be set")
+	}
+	if rs.cmd != nil {
+		return nil
+	}
+
+	rs.cmd = exec.Command(rs.cmdStr, rs.args...)
+	stderr, err := rs.cmd.StderrPipe()
+
+	go func() {
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/etcd_runner_stresser.go]: `defer close(rs.donec)` so `Close` only exits after the goroutine exits

File: tools/functional-tester/etcd-tester/etcd_runner_stresser.go [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+		out, err := ioutil.ReadAll(stderr)
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/etcd_runner_stresser.go]: will etcd-runner exit immediately after writing to stderr or is it possible this will keep reading never data from stderr and the runner never quits?

File: tools/functional-tester/etcd-tester/stresser.go [hunk 15]
```diff
@@ -149,6 +166,50 @@ func NewStresser(s string, sc *stressConfig, m *member) Stresser {
 			keysPerLease: sc.keysPerLease,
 			rateLimiter:  sc.rateLimiter,
 		}
+	case "election-runner":
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/stresser.go]: these should provision ops for themselves from `sc.rateLimiter` so they're not invisible to the rate limiting; probably via `SetLimit()`

File: tools/functional-tester/etcd-tester/stresser.go (hunk 8 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-tester/stresser.go]: sounds better.

File: tools/functional-tester/etcd-tester/etcd_runner_stresser.go [hunk 16]
```diff
@@ -0,0 +1,64 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os/exec"
+
```
Comment (gyuho) [inline tools/functional-tester/etcd-tester/etcd_runner_stresser.go]: remove empty line?

File: tools/functional-tester/etcd-runner/command/lease_renewer_command.go (hunk 5 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-runner/command/lease_renewer_command.go]: Since sigstop might stop runner for a long time in the case of compaction, lease will be expired if it's 5 seconds. I do agree a command line argument is better.

File: tools/functional-tester/etcd-runner/command/election_command.go (hunk 4 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-runner/command/election_command.go]: yep, when there are 3 elections running with the same elector string. it cause conflict with each other. the random string just make it rarely to happen. Maybe I should use pass in as a command arg? or something else to guarantee that elector string is unique.

File: pkg/stringutil/stringutil.go (hunk 1 above)
Comment (fanminshi) [inline pkg/stringutil/stringutil.go]: I agree.

File: tools/functional-tester/etcd-runner/command/lock_racer_command.go (hunk 6 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-runner/command/lock_racer_command.go]: agreed. Need to think a better approach.

File: tools/functional-tester/etcd-tester/stresser.go (hunk 7 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-tester/stresser.go]: Using `Pause` with `Stress` makes sense.

File: tools/functional-tester/etcd-tester/stresser.go (hunk 9 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-tester/stresser.go]: good point. will change runner to treat 0 as running forever.

File: tools/functional-tester/etcd-tester/etcd_runner_stresser.go (hunk 10 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-tester/etcd_runner_stresser.go]: yea, my bad. will change.

File: tools/functional-tester/etcd-tester/etcd_runner_stresser.go (hunk 11 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-tester/etcd_runner_stresser.go]: will change

File: tools/functional-tester/etcd-tester/etcd_runner_stresser.go (hunk 13 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-tester/etcd_runner_stresser.go]: yeah, forget about that. good catch!

File: tools/functional-tester/etcd-runner/command/election_command.go (hunk 4 above)
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/election_command.go]: Passing in an election name to etcd-runner would be good. It can be through args[0] in the command. The functional tester can name each instance with the current nano time.

File: tools/functional-tester/etcd-tester/etcd_runner_stresser.go (hunk 14 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-tester/etcd_runner_stresser.go]: if I understand runner code correctly, runner will exit on panic, log.panic, log.fataf, and `ExitWithError` which all of those will write to stderr and exit the program. it is possible that runner runs forever if it never encounters any problems. However, when the tester do clean up, tester will kill the runner which will close stderr.

Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/etcd_runner_stresser.go]: OK

File: tools/functional-tester/etcd-tester/etcd_runner_stresser.go (hunk 16 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-tester/etcd_runner_stresser.go]: good catch!

**Comment by fanminshi** [comment]
all fixe. PTAL

File: tools/functional-tester/etcd-runner/command/help.go [hunk 17]
```diff
@@ -14,7 +14,7 @@
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/help.go]: help.go should be in the same directory as the main.go, don't move this

File: tools/functional-tester/etcd-runner/command/lock_racer_command.go [hunk 18]
```diff
@@ -26,18 +26,24 @@ import (
 // NewLockRacerCommand returns the cobra command for "lock-racer runner".
 func NewLockRacerCommand() *cobra.Command {
 	cmd := &cobra.Command{
-		Use:   "lock-racer",
+		Use:   "lock-racer [name of lock (defaults to 'racers')]",
 		Short: "Performs lock race operation",
 		Run:   runRacerFunc,
 	}
-	cmd.Flags().IntVar(&rounds, "rounds", 100, "number of rounds to run")
+	cmd.Flags().IntVar(&rounds, "rounds", 100, "number of rounds to run. set rounds <= 0 to run forever")
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/lock_racer_command.go]: if `rounds` is already a command in root.go it shouldn't be redefined here

File: tools/functional-tester/etcd-runner/command/global.go [hunk 19]
```diff
@@ -61,16 +64,26 @@ func newClient(eps []string, timeout time.Duration) *clientv3.Client {
 	return c
 }
 
-func doRounds(rcs []roundClient, rounds int) {
-	var mu sync.Mutex
-	var wg sync.WaitGroup
+func doRounds(rcs []roundClient, rounds int, tokensPerRound int) {
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/global.go]: s/tokens/requests

File: tools/functional-tester/etcd-runner/command/global.go [hunk 20]
```diff
@@ -95,7 +108,11 @@ func doRounds(rcs []roundClient, rounds int) {
 	}
 
 	start := time.Now()
-	for i := 1; i < len(rcs)*rounds+1; i++ {
+	totalRounds := len(rcs)*rounds + 1
+	if rounds == maxRounds {
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/global.go]: remove this check

File: tools/functional-tester/etcd-runner/command/global.go [hunk 21]
```diff
@@ -95,7 +108,11 @@ func doRounds(rcs []roundClient, rounds int) {
...
+		totalRounds = rounds
+	}
+	for i := 1; i < totalRounds; i++ {
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/global.go]: i < totalResults || rounds <= 0

File: tools/functional-tester/etcd-runner/command/global.go [hunk 22]
```diff
@@ -61,16 +64,26 @@ func newClient(eps []string, timeout time.Duration) *clientv3.Client {
...
+		rounds = maxRounds
+	}
 
 	wg.Add(len(rcs))
 	finished := make(chan struct{})
+	limiter := rate.NewLimiter(rate.Limit(reqRate), reqRate)
 	for i := range rcs {
 		go func(rc *roundClient) {
 			defer wg.Done()
 			for rc.progress < rounds {
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/global.go]: rc.progress < rounds || rounds <= 0

File: tools/functional-tester/etcd-runner/command/global.go [hunk 23]
```diff
@@ -61,16 +64,26 @@ func newClient(eps []string, timeout time.Duration) *clientv3.Client {
...
-func doRounds(rcs []roundClient, rounds int) {
-	var mu sync.Mutex
-	var wg sync.WaitGroup
+func doRounds(rcs []roundClient, rounds int, tokensPerRound int) {
+	var (
+		mu sync.Mutex
+		wg sync.WaitGroup
+	)
+
+	if rounds < 1 {
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/global.go]: remove

File: tools/functional-tester/etcd-runner/command/global.go [hunk 24]
```diff
@@ -61,16 +64,26 @@ func newClient(eps []string, timeout time.Duration) *clientv3.Client {
 	return c
 }
 
-func doRounds(rcs []roundClient, rounds int) {
-	var mu sync.Mutex
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/global.go]: keep this; the new version just eats more lines

File: tools/functional-tester/etcd-runner/command/global.go [hunk 25]
```diff
@@ -23,24 +24,26 @@ import (
...
+	reqRate     int // maximum number of requests per second.
 )
 
-// GlobalFlags are flags that defined globally
-// and are inherited to all sub-commands.
-type GlobalFlags struct {
-	Endpoints   []string
-	DialTimeout time.Duration
-}
+const maxRounds = 1<<31 - 1
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/global.go]: remove

File: tools/functional-tester/etcd-runner/command/lock_racer_command.go [hunk 26]
```diff
@@ -26,18 +26,24 @@ import (
...
 	cmd.Flags().IntVar(&totalClientConnections, "total-client-connections", 10, "total number of client connections")
 	return cmd
 }
 
 func runRacerFunc(cmd *cobra.Command, args []string) {
-	if len(args) > 0 {
-		ExitWithError(ExitBadArgs, errors.New("lock-racer does not take any argument"))
+	racers := "racers"
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/lock_racer_command.go]: move this directly above `len(args) == 1`; otherwise it's splitting up the `racers` setting logic for no reason

File: tools/functional-tester/etcd-runner/command/global.go [hunk 27]
```diff
@@ -23,24 +24,26 @@ import (
...
+	"golang.org/x/time/rate"
 )
 
 var (
-	rounds                 int           // total number of rounds the operation needs to be performed
 	totalClientConnections int           // total number of client connections to be made with server
 	noOfPrefixes           int           // total number of prefixes which will be watched upon
 	watchPerPrefix         int           // number of watchers per prefix
-	reqRate                int           // put request per second
+	prefix                 string        // prefix append to keys in watcher
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/global.go]: this should be `watchPrefix` in `watch_command.go`

File: tools/functional-tester/etcd-runner/command/global.go [hunk 28]
```diff
@@ -23,24 +24,26 @@ import (
...
 	totalKeys              int           // total number of keys for operation
 	runningTime            time.Duration // time for which operation should be performed
+	ttl                    int           // lease ttl in lease renewer
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/global.go]: this should be `leaseTTL` in `lease_renewer_command.go` if it's only being used by one command

File: tools/functional-tester/etcd-runner/command/global.go [hunk 29]
```diff
@@ -23,24 +24,26 @@ import (
...
+
+	// global flags
+	endpoints   []string
+	dialTimeout time.Duration
+	rounds      int // total number of rounds the operation needs to be performed and set it to <= 0 to run maxRounds.
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/global.go]: // total number of rounds to run; set to <= 0 to run forever

File: tools/functional-tester/etcd-runner/command/global.go [hunk 30]
```diff
@@ -95,7 +108,11 @@ func doRounds(rcs []roundClient, rounds int) {
 	}
 
 	start := time.Now()
-	for i := 1; i < len(rcs)*rounds+1; i++ {
+	totalRounds := len(rcs)*rounds + 1
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/global.go]: Don't call this `totalRounds`. That value is already `rounds`. Call it `totalResults` since it's counting the total results for all rounds.

File: tools/functional-tester/etcd-runner/command/global.go [hunk 31]
```diff
@@ -61,16 +64,26 @@ func newClient(eps []string, timeout time.Duration) *clientv3.Client {
...
+	}
 
 	wg.Add(len(rcs))
 	finished := make(chan struct{})
+	limiter := rate.NewLimiter(rate.Limit(reqRate), reqRate)
 	for i := range rcs {
 		go func(rc *roundClient) {
 			defer wg.Done()
 			for rc.progress < rounds {
+				if err := limiter.WaitN(context.Background(), tokensPerRound); err != nil {
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/global.go]: I think this should be `tokensPerRound/len(rcs)` since there can be several roundClients working on a round

File: tools/functional-tester/etcd-runner/command/lease_renewer_command.go [hunk 32]
```diff
@@ -34,14 +35,14 @@ func NewLeaseRenewerCommand() *cobra.Command {
...
 	}
+	cmd.Flags().IntVar(&ttl, "ttl", 5, "lease's ttl")
 	return cmd
 }
 
 func runLeaseRenewerFunc(cmd *cobra.Command, args []string) {
 	if len(args) > 0 {
 		ExitWithError(ExitBadArgs, errors.New("lease-renewer does not take any argument"))
 	}
-
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/lease_renewer_command.go]: put newline back?

File: tools/functional-tester/etcd-runner/command/root.go [hunk 33]
```diff
@@ -0,0 +1,69 @@
...
+	cobra.EnablePrefixMatching = true
+
+	rand.Seed(time.Now().UnixNano())
+
+	log.SetFlags(log.Lmicroseconds)
+
+	rootCmd.PersistentFlags().StringSliceVar(&endpoints, "endpoints", []string{"127.0.0.1:2379"}, "gRPC endpoints")
+	rootCmd.PersistentFlags().DurationVar(&dialTimeout, "dial-timeout", defaultDialTimeout, "dial timeout for client connections")
+	rootCmd.Flags().IntVar(&reqRate, "req-rate", 30, "maximum number of requests per second")
+	rootCmd.Flags().IntVar(&rounds, "rounds", 100, "number of rounds to run")
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/root.go]: `number of rounds to run; 0 to run forever`

File: tools/functional-tester/etcd-runner/command/watch_command.go [hunk 34]
```diff
@@ -50,6 +51,9 @@ func runWatcherFunc(cmd *cobra.Command, args []string) {
 	}
 
 	ctx := context.Background()
+	if rounds <= 0 {
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/watch_command.go]: remove

File: tools/functional-tester/etcd-runner/command/watch_command.go [hunk 35]
```diff
@@ -50,6 +51,9 @@ func runWatcherFunc(cmd *cobra.Command, args []string) {
...
+		rounds = maxRounds
+	}
 	for round := 0; round < rounds; round++ {
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/watch_command.go]: round < rounds || rounds <= 0

File: tools/functional-tester/etcd-runner/command/root.go [hunk 36]
```diff
@@ -0,0 +1,69 @@
+// Copyright 2017 The etcd Authors
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/root.go]: please keep this in `main.go` to keep the diff short

File: tools/functional-tester/etcd-tester/stresser.go [hunk 37]
```diff
@@ -28,8 +29,10 @@ func init() { grpclog.SetLogger(plog) }
 type Stresser interface {
 	// Stress starts to stress the etcd cluster
 	Stress() error
-	// Cancel cancels the stress test on the etcd cluster
-	Cancel()
+	// Puase pauses the stresser on stressing the etcd cluster
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/stresser.go]: `Pause stops the stresser from sending requests to etcd. Resume by calling Stress.`

File: tools/functional-tester/etcd-tester/stresser.go [hunk 38]
```diff
@@ -28,8 +29,10 @@ func init() { grpclog.SetLogger(plog) }
...
+	Pause()
+	// Close releases resources of Stresser
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/stresser.go]: `Close releases all of the Stresser's resources.`

File: tools/functional-tester/etcd-tester/tester.go [hunk 39]
```diff
@@ -229,19 +229,25 @@ func (tt *tester) cleanup() error {
...
-	tt.stresser.Cancel()
-	plog.Infof("%s canceled stressers", tt.logPrefix())
+func (tt *tester) pauseStresser() {
+	plog.Infof("%s pausing the stressers...", tt.logPrefix())
+	tt.stresser.Pause()
+	plog.Infof("%s paused stressers", tt.logPrefix())
 }
 
-func (tt *tester) startStresser() (err error) {
+func (tt *tester) stressStresser() (err error) {
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/tester.go]: keep this named `startStresser`; `stressStresser` is awkward

File: tools/functional-tester/etcd-tester/stresser.go [hunk 40]
```diff
@@ -149,6 +167,68 @@ func NewStresser(s string, sc *stressConfig, m *member) Stresser {
...
+				"--rounds=0", // runs forever
+				"--req-rate", fmt.Sprintf("%v", reqRate),
+			},
+			errc:  make(chan error, 1),
+			donec: make(chan struct{}),
+		}
+	case "watch-runner":
+		reqRate := 100
+		sc.rateLimiter.SetLimit(sc.rateLimiter.Limit() - rate.Limit(reqRate))
+		return &runnerStresser{
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/stresser.go]: this should have a constructor `newRunnerStresser(args, sc, reqRate)` since there's repeated setup for each runner command

File: tools/functional-tester/etcd-tester/key_stresser.go [hunk 41]
```diff
@@ -140,13 +140,17 @@ func (s *keyStresser) run(ctx context.Context) {
...
-func (s *keyStresser) Cancel() {
+func (s *keyStresser) Pause() {
 	s.cancel()
 	s.conn.Close()
 	s.wg.Wait()
 	plog.Infof("keyStresser %q is canceled", s.Endpoint)
 }
 
+func (s *keyStresser) Close() {
+	s.Pause()
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/key_stresser.go]: This is backwards since Pause is calling conn.Close. Put the Pause code in here and have Pause call Close instead

File: tools/functional-tester/etcd-tester/lease_stresser.go [hunk 42]
```diff
@@ -361,13 +361,17 @@ func (ls *leaseStresser) randomlyDropLease(leaseID int64) (bool, error) {
 	return false, ls.ctx.Err()
 }
 
-func (ls *leaseStresser) Cancel() {
-	plog.Debugf("lease stresser %q is canceling...", ls.endpoint)
+func (ls *leaseStresser) Pause() {
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/lease_stresser.go]: s/Pause/Close since this is closing connections

File: tools/functional-tester/etcd-tester/lease_stresser.go [hunk 43]
```diff
@@ -361,13 +361,17 @@ func (ls *leaseStresser) randomlyDropLease(leaseID int64) (bool, error) {
...
+	plog.Debugf("lease stresser %q is pausing...", ls.endpoint)
 	ls.cancel()
 	ls.runWg.Wait()
 	ls.aliveWg.Wait()
 	ls.conn.Close()
-	plog.Infof("lease stresser %q is canceled", ls.endpoint)
+	plog.Infof("lease stresser %q is paused", ls.endpoint)
+}
+
+func (ls *leaseStresser) Close() {
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/lease_stresser.go]: s/Close/Pause

File: tools/functional-tester/etcd-tester/etcd_runner_stresser.go [hunk 44]
```diff
@@ -0,0 +1,63 @@
...
+func (rs *runnerStresser) Pause() {
+	syscall.Kill(rs.cmd.Process.Pid, syscall.SIGSTOP)
+}
+
+func (rs *runnerStresser) Close() {
+	syscall.Kill(rs.cmd.Process.Pid, syscall.SIGINT)
+	rs.cmd.Wait()
+	<-rs.donec
+}
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/etcd_runner_stresser.go]: this should add the reqRate back to the ratelimiter

File: tools/functional-tester/etcd-runner/command/root.go (hunk 36 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-runner/command/root.go]: I am following the convention from cobra doc. https://github.com/spf13/cobra#manually-create-rootcmd.

let the rootcmd to live in the same pkg as rest of cmds allows a simpler way of sharing global variable, or else the sub cmd needs to access the etcd-runner pkg to get it, https://github.com/coreos/etcd/blob/master/tools/functional-tester/etcd-runner/main.go#L34.

File: tools/functional-tester/etcd-runner/command/global.go (hunk 24 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-runner/command/global.go]: var() block always eat more lines. In what case should we use it then?

File: tools/functional-tester/etcd-runner/command/global.go (hunk 31 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-runner/command/global.go]: agreed.

File: tools/functional-tester/etcd-tester/tester.go (hunk 39 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-tester/tester.go]: good catch.

**Comment by fanminshi** [comment]
The current code experience error from election-runner and watch-runner.
Investigating...

```
watcher-runner error

19:45:54 stresser | (bin/etcd-runner [watcher --prefix 1491533067143843386 --total-keys=1 --total-prefixes=1 --watch-per-prefix=1 --endpoints 127.0.0.1:12379 --rounds=0 --req-rate 100]) stderr 19:45:53.539720 expect more keys (received 1/0) for 1491533067143843386 0-v8r4j

election-runner error

19:51:10 stresser | etcd-tester: [round#0 case#9] (bin/etcd-runner [election 1491533403001749852 --dial-timeout=10s --endpoints 127.0.0.1:32379 --total-client-connections=10 --rounds=0 --req-rate 100]) stderr 19:51:06.765943 no progress after 1 minute!
19:51:10 stresser | panic: no progress after 1 minute!
19:51:10 stresser | goroutine 1 [running]:
19:51:10 stresser | log.Panic(0xc4204f5ae0, 0x1, 0x1)
19:51:10 stresser | 	/usr/local/go/src/log/log.go:322 +0xc0
19:51:10 stresser | github.com/coreos/etcd/tools/functional-tester/etcd-runner/command.doRounds(0xc42005e340, 0xa, 0xa, 0x0, 0x14)
19:51:10 stresser | 	/Users/fanminshi/work/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/command/global.go:106 +0x4e1
19:51:10 stresser | github.com/coreos/etcd/tools/functional-tester/etcd-runner/command.runElectionFunc(0xc420182000, 0xc420184100, 0x1, 0x8)
19:51:10 stresser | 	/Users/fanminshi/work/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/command/election_command.go:130 +0x7c0
19:51:10 stresser | github.com/spf13/cobra.(*Command).execute(0xc420182000, 0xc420184080, 0x8, 0x8, 0xc420182000, 0xc420184080)
19:51:10 stresser | 	/Users/fanminshi/work/src/github.com/coreos/etcd/gopath/src/github.com/spf13/cobra/command.go:572 +0x203
19:51:10 stresser | github.com/spf13/cobra.(*Command).ExecuteC(0x1a51680, 0x0, 0xc420141290, 0xc4200c7f58)
19:51:10 stresser | 	/Users/fanminshi/work/src/github.com/coreos/etcd/gopath/src/github.com/spf13/cobra/command.go:662 +0x394
19:51:10 stresser | github.com/spf13/cobra.(*Command).Execute(0x1a51680, 0x173d357, 0x7)
19:51:10 stresser | 	/Users/fanminshi/work/src/github.com/coreos/etcd/gopath/src/github.com/spf13/cobra/command.go:618 +0x2b
19:51:10 stresser | github.com/coreos/etcd/tools/functional-tester/etcd-runner/command.Start()
19:51:10 stresser | 	/Users/fanminshi/work/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/command/root.go:66 +0x8e
19:51:10 stresser | main.main()
19:51:10 stresser | 	/Users/fanminshi/work/src/github.com/coreos/etcd/cmd/tools/functional-tester/etcd-runner/main.go:21 +0x20
```

**Comment by fanminshi** [comment]
I separate out the runner logic to this pr, https://github.com/coreos/etcd/pull/7812.

**Comment by fanminshi** [comment]
All fixed. PTAL

**Comment by heyitsanthony** [comment]
lgtm
